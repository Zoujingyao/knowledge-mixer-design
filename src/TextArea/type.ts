import React, { ReactNode } from 'react';

export interface TextAreaProps {
  /**
   * @description 类名
   */
  className?: string;
  /**
   * @description 自定义样式
   * @default {}
   */
  moreStyle?: object;
  /**
   * @description 提示
   * @default ''
   */
  placeholder?: string;
  /**
   * @description 文本域内容最大长度
   */
  maxLength?: number;
  /**
   * @description 输入框内容改变回调
   * @params e
   */
  handleIptChange?: (e: string) => void;
  /**
   * @description 输入框聚焦回调
   */
  handleIptFocus?: (e: React.FocusEvent<HTMLTextAreaElement>) => void;
  /**
   * @description 输入框点击回调
   */
  handleClick?: (e: React.MouseEvent<HTMLTextAreaElement>) => void;
  /**
   * @description 输入框失去焦点回调
   */
  handleIptBlur?: (e: React.FocusEvent<HTMLTextAreaElement>) => void;
  /**
   * @description 输入框键盘监听
   */
  handleKeyDown?: (e: React.KeyboardEvent<HTMLTextAreaElement>) => void;
  /**
   * @description 默认内容
   * @default ''
   */
  defaultValue?: string;
}
