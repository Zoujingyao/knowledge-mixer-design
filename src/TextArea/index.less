@gray-color: #c5c5c5;
.textarea {
  display: -webkit-box;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-sizing: border-box;
  box-sizing: border-box;
  width: 100%;
  height: 56px;
  min-height: 56px;
  padding: 5px 8px;

  font-size: 14px;
  font-family: PingFangSC-Regular;
  line-height: 22px;
  border: 1px solid #d9d9d9;
  border-radius: 5pxs;
  outline: none;
  -webkit-transition: all 0.3s cubic-bezier(0.38, 0, 0.24, 1), height 0s;
  transition: all 0.3s cubic-bezier(0.38, 0, 0.24, 1), height 0s;
  resize: vertical;
  &:hover {
    border-color: #325dff;
  }
  &:focus {
    border-color: #b1c2ff;
    box-shadow: 0 0 5px 1px #b8c7ff;
  }
}
.textarea-limit {
  position: absolute;
  right: 16px;
  bottom: 8px;
  color: @gray-color;
  font-size: 14px;
}
.clear-svg {
  position: absolute;
  top: 5px;
  right: 0;
  padding: 0 5px;
  color: #00000040;
  font-size: 12px;
  cursor: pointer;
}
