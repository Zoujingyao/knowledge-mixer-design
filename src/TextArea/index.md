---
title: TextArea 文本输入域
nav:
  title: Components 
  path: /components 
group:
  title: 数据录入
  path: /components/data-input
---

# TextArea 文本输入域

## 代码演示

### 基本
```tsx
/**
 * description: 用于多行输入。
 */
import React from 'react';
import { TextArea } from 'knowledge-mixer-design';
export default ()=>{
  return <TextArea/>
}
```

<API src="./index.tsx"></API>
