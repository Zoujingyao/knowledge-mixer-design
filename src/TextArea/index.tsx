import React, { FC, PropsWithChildren, useContext, useEffect, useState } from 'react';
import { TextAreaProps } from './type';
import './index.less';
const TextArea: FC<PropsWithChildren<TextAreaProps>> = (props) => {
  const {
    moreStyle,
    placeholder,
    maxLength,
    handleIptChange,
    handleKeyDown,
    handleIptFocus,
    handleClick,
    handleIptBlur,
    defaultValue,
  } = props;
  const [txaValue, setTxaValue] = useState<string>(defaultValue || '');
  const [isComposition, setComposition] = useState<boolean>(false);
  useEffect(() => {
    // eslint-disable-next-line @typescript-eslint/no-unused-expressions
    handleIptChange && handleIptChange(txaValue);
  // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [txaValue]);
  const onChangeTxa = (e: React.ChangeEvent<HTMLTextAreaElement>) => {
    if (maxLength && !isComposition && maxLength < e.target.value.length) {
      setTxaValue(e.target.value.slice(0, maxLength));
    } else {
      setTxaValue(e.target.value);
    }
  };
  const onCompositionStartTxa = (e: React.CompositionEvent<HTMLTextAreaElement>) => {
    setComposition(true);
  };
  const onCompositionEndTxa = (e: React.CompositionEvent<HTMLTextAreaElement>) => {
    setComposition(false);
    if (maxLength) {
      setTxaValue(txaValue.slice(0, maxLength));
    }
  };

  const blurTxa = (e: React.FocusEvent<HTMLTextAreaElement>) => {
    // eslint-disable-next-line @typescript-eslint/no-unused-expressions
    handleIptBlur && handleIptBlur(e);
  };
  const focusTxa = (e: React.FocusEvent<HTMLTextAreaElement>) => {
    // eslint-disable-next-line @typescript-eslint/no-unused-expressions
    handleIptFocus && handleIptFocus(e);
  };
  const txaHandleClick = (e: React.MouseEvent<HTMLTextAreaElement>) => {
    // eslint-disable-next-line @typescript-eslint/no-unused-expressions
    handleClick && handleClick(e);
  };
  return (
    <div>
      <textarea
        className="textarea"
        style={moreStyle}
        placeholder={placeholder}
        value={defaultValue || txaValue}
        onChange={onChangeTxa}
        onBlur={blurTxa}
        onFocus={focusTxa}
        onKeyUp={(e) => handleKeyDown && handleKeyDown(e)}
        onClick={txaHandleClick}
        onCompositionStart={onCompositionStartTxa}
        onCompositionEnd={onCompositionEndTxa}
       />
    </div>
  );
};
export default React.memo(TextArea);
