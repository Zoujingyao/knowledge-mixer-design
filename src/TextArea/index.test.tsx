import React from 'react';
import renderer from 'react-test-renderer';
import '@testing-library/jest-dom';
import TextArea from './index';

describe('<TextArea />', () => {
  it('render TextArea with dumi', () => {
    try {
      const ele = renderer.create(<TextArea />).toJSON();
      expect(ele).toMatchSnapshot();
    } catch (e) {
      console.log(e);
    }
  });
});
