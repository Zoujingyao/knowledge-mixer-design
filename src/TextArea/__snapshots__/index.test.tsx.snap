// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextArea /> render TextArea with dumi 1`] = `
<div>
  <textarea
    className="textarea"
    onBlur={[Function]}
    onChange={[Function]}
    onClick={[Function]}
    onCompositionEnd={[Function]}
    onCompositionStart={[Function]}
    onFocus={[Function]}
    onKeyUp={[Function]}
    value=""
  />
</div>
`;
